src/nvapi/nvapi_d3d12_device.cpp

```diff
@@ -27,180 +91,86 @@ namespace dxvk {
+        return SUCCEEDED(m_vkd3dDevice->GetCudaTextureObject(srvHandle, samplerHandle, reinterpret_cast<UINT32*>(cudaTextureHandle)));
     }
 
-    bool NvapiD3d12Device::GetCudaSurfaceObject(ID3D12Device* device, D3D12_CPU_DESCRIPTOR_HANDLE uavHandle, NvU32* cudaSurfaceHandle) {
-        auto cubinDevice = GetCubinDevice(device);
-        if (cubinDevice == nullptr)
+    bool NvapiD3d12Device::GetCudaSurfaceObject(D3D12_CPU_DESCRIPTOR_HANDLE uavHandle, NvU32* cudaSurfaceHandle) const {
+        if (!m_vkd3dDevice || !m_supportsNvxBinaryImport)
             return false;
 
-        return SUCCEEDED(cubinDevice->GetCudaSurfaceObject(uavHandle, reinterpret_cast<UINT32*>(cudaSurfaceHandle)));
+        return SUCCEEDED(m_vkd3dDevice->GetCudaSurfaceObject(uavHandle, reinterpret_cast<UINT32*>(cudaSurfaceHandle)));
     }
 
-    bool NvapiD3d12Device::NotifyOutOfBandCommandQueue(ID3D12CommandQueue* commandQueue, D3D12_OUT_OF_BAND_CQ_TYPE type) {
-        auto commandQueueExt = GetCommandQueueExt(commandQueue);
-        if (commandQueueExt == nullptr)
+    bool NvapiD3d12Device::NotifyOutOfBandCommandQueue(D3D12_OUT_OF_BAND_CQ_TYPE type) const {
+        if (!m_vkd3dCommandQueue)
             return false;
 
-        return SUCCEEDED(commandQueueExt->NotifyOutOfBandCommandQueue(type));
+        return SUCCEEDED(m_vkd3dCommandQueue->NotifyOutOfBandCommandQueue(type));
     }
 
-    bool NvapiD3d12Device::LaunchCubinShader(ID3D12GraphicsCommandList* commandList, NVDX_ObjectHandle pShader, NvU32 blockX, NvU32 blockY, NvU32 blockZ, const void* params, NvU32 paramSize) {
-        auto commandListExt = GetCommandListExt(commandList);
-        if (!commandListExt.has_value())
+    bool NvapiD3d12Device::LaunchCubinShader(NVDX_ObjectHandle pShader, NvU32 blockX, NvU32 blockY, NvU32 blockZ, const void* params, NvU32 paramSize) const {
+        if (!m_vkd3dGraphicsCommandList)
             return false;
 
-        auto cmdList = commandListExt.value().CommandListExt;
-        auto interfaceVersion = commandListExt.value().InterfaceVersion;
-
         uint32_t smem = 0;
-        std::scoped_lock lock(m_cubinSmemMutex);
-        auto it = m_cubinSmemMap.find(pShader);
-        if (it != m_cubinSmemMap.end())
-            smem = it->second;
-        else
-            log::info("Failed to find CuBIN in m_cubinSmemMap, defaulting to 0");
-
-        if (interfaceVersion >= 1)
-            return SUCCEEDED(cmdList->LaunchCubinShaderEx(reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(pShader), blockX, blockY, blockZ, smem, params, paramSize, nullptr, 0));
-        else {
-            if (smem != 0)
-                log::info("Non-zero SMEM value supplied for CuBIN but ID3D12GraphicsCommandListExt1 not supported! This may cause corruption");
-
-            return SUCCEEDED(cmdList->LaunchCubinShader(reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(pShader), blockX, blockY, blockZ, params, paramSize));
+        if (m_supportsExtGraphicsCommandList1 || log::tracing()) {
```

Comment:
Why `|| log::tracing()` here? I'd expect something like this shouldn't change the observable behavior… or cause misleading messages to be reported because the underlying implementation doesn't support GCLExt1 even if the application handles it gracefully.

Reply:
The idea is to avoid the lock and map lookup when Ext0 is used. When not using Ext1, `smem` is only used for the log statement below. So only logging that statement on tracing, let's us skip the lock and lookup.

Reply:
I've restored the log message, but still kept the lock scoped.